docs/articles/streams/error-handling.md

```diff
@@ -1,78 +1,155 @@
 ---
-layout: docs.hbs
-title: Error Handling
+uid: error-handling
+title: Error Handling in Streams
 ---
+# Error Handling in Streams
 
-Strategies for how to handle exceptions from processing stream elements can be defined when
-materializing the stream. The error handling strategies are inspired by actor supervision
-strategies, but the semantics have been adapted to the domain of stream processing.
+When a stage in a stream fails this will normally lead to the entire stream being torn down. Each of the stages downstream gets informed about the failure and each upstream stage sees a cancellation.
 
-> [!WARNING]
-> *ZipWith*, *GraphStage* junction, *ActorPublisher* source and *ActorSubscriber* sink 
-components do not honour the supervision strategy attribute yet.
+In many cases you may want to avoid complete stream failure, this can be done in a few different ways:
 
-# Supervision Strategies
+- `Recover` to emit a final element then complete the stream normally on upstream failure
+- `RecoverWithRetries` to create a new upstream and start consuming from that on failure
+- Using a supervision strategy for stages that support it
 
-There are three ways to handle exceptions from application code:
+In addition to these built in tools for error handling, a common pattern is to wrap the stream inside an actor, and have the actor restart the entire stream on failure.
+
+## Recover
+`Recover` allows you to emit a final element and then complete the stream on an upstream failure. Deciding which exceptions should be recovered is done through a `delegate`. If an exception does not have a matching case the stream is failed.
+
+Recovering can be useful if you want to gracefully complete a stream on failure while letting downstream know that there was a failure.
+
+```C#
+Source.From(Enumerable.Range(0, 6)).Select(n =>
+    {
+        if (n < 5)
+            return n.ToString();
+
+        throw new ArithmeticException("Boom!");
+    })
+    .Recover(exception =>
+    {
+        if (exception is ArithmeticException)
+            return new Option<string>("stream truncated");
+        return Option<string>.None;
```

Comment:
@Silv3rcircl3 I wondered that we have such API. Why not just string?

Reply:
`Option<>` is like `Nullable<>`, but it can also work with non-value types. We needed a universal primitive that will notify about value absence and will work with both (classes and structs).

Reply:
Also note that NULL is not allowed as element, so return null would fail the stream.

Reply:
@Silv3rcircl3 Is null is accepted in `RecoveryWithRetries`?

Reply:
Yup, because there you return a complete new source and null indicates that you don't want to recover but rather fail the stream. By element I mean a single message that is processed by the stream, i.e. Select(x => null as string) would fail the stream

Reply:
Ok, thanks